File: clients/macos/Tests/CoreTests/DesktopReleaseCatalogTests.swift
import Foundation
import XCTest
@testable import HeteroNetworkCore

final class DesktopReleaseCatalogTests: XCTestCase {
    private let assetName = "heteronetwork-client-macos-arm64.zip"

    func testSelectsNewestCompletePublishedRelease() throws {
        let update = try DesktopReleaseCatalog.availableUpdate(
            from: catalog([
                release(tag: "v0.1.15-dev.18", publishedAt: "2026-09-19T12:00:00Z", assets: []),
                release(tag: "v0.1.15-dev.17", publishedAt: "2026-09-19T11:00:00Z"),
                release(tag: "v0.1.15-dev.16", publishedAt: "2026-09-19T10:00:00Z"),
            ]),
            currentTag: "v0.1.15-dev.16",
            assetName: assetName
        )

        XCTAssertEqual(update?.tag, "v0.1.15-dev.17")
        XCTAssertEqual(update?.assetName, assetName)
    }

    func testReportsCurrentWhenNewestCompleteReleaseIsInstalled() throws {
        let update = try DesktopReleaseCatalog.availableUpdate(
            from: catalog([
                release(tag: "v0.1.15-dev.17", publishedAt: "2026-09-19T11:00:00Z"),
                release(tag: "v0.1.15-dev.16", publishedAt: "2026-09-19T10:00:00Z"),
            ]),
            currentTag: "v0.1.15-dev.17",
            assetName: assetName
        )

        XCTAssertNil(update)
    }

    func testDoesNotDowngradeWhenCurrentReleaseAssetsWereWithdrawn() throws {
        let update = try DesktopReleaseCatalog.availableUpdate(
            from: catalog([
                release(tag: "v0.1.15-dev.18", publishedAt: "2026-09-19T12:00:00Z", assets: []),
                release(tag: "v0.1.15-dev.17", publishedAt: "2026-09-19T11:00:00Z"),
            ]),
            currentTag: "v0.1.15-dev.18",
            assetName: assetName
        )

        XCTAssertNil(update)
    }

    func testIgnoresDraftAndUntrustedDownloadURLs() throws {
        var draft = release(tag: "v0.1.15-dev.19", publishedAt: "2026-09-19T13:00:00Z")
        draft["draft"] = true
        var untrusted = release(tag: "v0.1.15-dev.18", publishedAt: "2026-09-19T12:00:00Z")
        untrusted["assets"] = [
            ["name": assetName, "browser_download_url": "https://example.com/client.zip"],
            ["name": "\(assetName).sha256", "browser_download_url": downloadURL(
                tag: "v0.1.15-dev.18",
                name: "\(assetName).sha256"
            )],
        ]

        let update = try DesktopReleaseCatalog.availableUpdate(
            from: catalog([
                draft,
                untrusted,
                release(tag: "v0.1.15-dev.17", publishedAt: "2026-09-19T11:00:00Z"),
            ]),
            currentTag: "v0.1.15-dev.16",
            assetName: assetName
        )

        XCTAssertEqual(update?.tag, "v0.1.15-dev.17")
    }

    func testDevelopmentBuildDoesNotOfferReleaseUpdate() throws {
        let update = try DesktopReleaseCatalog.availableUpdate(
            from: catalog([
                release(tag: "v0.1.15-dev.17", publishedAt: "2026-09-19T11:00:00Z"),
            ]),
            currentTag: "development",
            assetName: assetName
        )

        XCTAssertNil(update)
    }

    func testAtomFeedSelectsNewestReleaseAndBuildsTrustedURLs() throws {
        let update = try DesktopReleaseCatalog.availableUpdate(
            fromAtom: atomFeed([
                atomEntry(tag: "v0.1.15-dev.24", updated: "2026-09-21T08:02:47Z"),
                atomEntry(tag: "v0.1.15-dev.23", updated: "2026-09-21T07:02:47Z"),
            ]),
            currentTag: "v0.1.15-dev.23",
            assetName: assetName
        )

        XCTAssertEqual(update?.tag, "v0.1.15-dev.24")
        XCTAssertEqual(update?.archiveURL.absoluteString, downloadURL(
            tag: "v0.1.15-dev.24",
            name: assetName
        ))
        XCTAssertEqual(update?.checksumURL.absoluteString, downloadURL(
            tag: "v0.1.15-dev.24",
            name: "\(assetName).sha256"
        ))
    }

    func testAtomFeedDoesNotOfferCurrentOrOlderRelease() throws {
        let current = try DesktopReleaseCatalog.availableUpdate(
            fromAtom: atomFeed([
                atomEntry(tag: "v0.1.15-dev.24", updated: "2026-09-21T08:02:47Z"),
                atomEntry(tag: "v0.1.15-dev.23", updated: "2026-09-21T07:02:47Z"),
            ]),
            currentTag: "v0.1.15-dev.24",
            assetName: assetName
        )
        XCTAssertNil(current)

        let newerInstalled = try DesktopReleaseCatalog.availableUpdate(
            fromAtom: atomFeed([
                atomEntry(tag: "v0.1.15-dev.25", updated: "2026-09-21T09:02:47Z"),
                atomEntry(tag: "v0.1.15-dev.24", updated: "2026-09-21T08:02:47Z"),
            ]),
            currentTag: "v0.1.15-dev.25",
            assetName: assetName
        )
        XCTAssertNil(newerInstalled)
    }

    func testAtomFeedIgnoresUntrustedAndUnsafeReleaseLinks() throws {
        let update = try DesktopReleaseCatalog.availableUpdate(
            fromAtom: atomFeed([
                atomEntry(
                    tag: "v0.1.15-dev.25",
                    updated: "2026-09-21T09:02:47Z",
                    href: "https://example.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/tag/v0.1.15-dev.25"
                ),
                atomEntry(
                    tag: "v0.1.15-dev.24",
                    updated: "2026-09-21T08:02:47Z",
                    href: "https://github.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/tag/v0.1.15-dev.24%2Fescape"
                ),
                atomEntry(tag: "v0.1.15-dev.23", updated: "2026-09-21T07:02:47Z"),
            ]),
            currentTag: "v0.1.15-dev.22",
            assetName: assetName
        )

        XCTAssertEqual(update?.tag, "v0.1.15-dev.23")
    }

    func testRejectsMalformedAtomFeed() {
        XCTAssertThrowsError(
            try DesktopReleaseCatalog.availableUpdate(
                fromAtom: Data("<feed><entry>".utf8),
                currentTag: "v0.1.15-dev.23",
                assetName: assetName
            )
        )
    }

    private func release(
        tag: String,
        publishedAt: String,
        assets: [[String: Any]]? = nil
    ) -> [String: Any] {
        [
            "tag_name": tag,
            "published_at": publishedAt,
            "draft": false,
            "assets": assets ?? [
                ["name": assetName, "browser_download_url": downloadURL(tag: tag, name: assetName)],
                [
                    "name": "\(assetName).sha256",
                    "browser_download_url": downloadURL(tag: tag, name: "\(assetName).sha256"),
                ],
            ],
        ]
    }

    private func downloadURL(tag: String, name: String) -> String {
        "https://github.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/download/\(tag)/\(name)"
    }

    private func catalog(_ releases: [[String: Any]]) throws -> Data {
        try JSONSerialization.data(withJSONObject: releases, options: [.sortedKeys])
    }

    private func atomEntry(
        tag: String,
        updated: String,
        href: String? = nil
    ) -> String {
        let releaseURL = href
            ?? "https://github.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/tag/\(tag)"
        return """
        <entry>
          <updated>\(updated)</updated>
          <link rel="alternate" type="text/html" href="\(releaseURL)" />
          <title>HeteroNetwork \(tag)</title>
        </entry>
        """
    }

    private func atomFeed(_ entries: [String]) -> Data {
        Data("""
        <?xml version="1.0" encoding="UTF-8"?>
        <feed xmlns="http://www.w3.org/2005/Atom">
        \(entries.joined(separator: "\n"))
        </feed>
        """.utf8)
    }
}

File: clients/windows/tests/HeteroNetwork.Core.Tests/DesktopReleaseCatalogTests.cs
using System.Text;
using System.Text.Json;
using HeteroNetwork.Core;

namespace HeteroNetwork.Core.Tests;

public sealed class DesktopReleaseCatalogTests
{
    private const string Asset = DesktopReleaseCatalog.WindowsAssetName;

    [Fact]
    public void SelectsNewestCompletePublishedRelease()
    {
        var catalog = Catalog(
            Release("v1.2.0", "2026-09-20T00:00:00Z", assets: [Asset, $"{Asset}.sha256"]),
            Release("v1.3.0", "2026-09-21T00:00:00Z", draft: true, assets: [Asset, $"{Asset}.sha256"]),
            Release("v1.1.0", "2026-09-19T00:00:00Z", assets: [Asset, $"{Asset}.sha256"]));

        var update = Assert.IsType<DesktopReleaseUpdate>(
            DesktopReleaseCatalog.AvailableUpdate(catalog, "v1.1.0"));

        Assert.Equal("v1.2.0", update.Tag);
        Assert.Equal(Asset, update.AssetName);
        Assert.Equal(
            $"https://github.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/download/v1.2.0/{Asset}",
            update.ArchiveUrl.AbsoluteUri);
    }

    [Fact]
    public void DoesNotDowngradeAReleaseWithNewerPublicationTime()
    {
        var catalog = Catalog(
            Release("v2.0.0", "2026-09-21T00:00:00Z", assets: []),
            Release("v1.9.0", "2026-09-20T00:00:00Z", assets: [Asset, $"{Asset}.sha256"]));

        Assert.Null(DesktopReleaseCatalog.AvailableUpdate(catalog, "v2.0.0"));
    }

    [Fact]
    public void RejectsAssetsHostedOutsideTheExpectedReleasePath()
    {
        var release = new
        {
            tag_name = "v1.2.0",
            published_at = "2026-09-20T00:00:00Z",
            draft = false,
            assets = new[]
            {
                new
                {
                    name = Asset,
                    browser_download_url = $"https://example.com/{Asset}",
                },
                new
                {
                    name = $"{Asset}.sha256",
                    browser_download_url = DownloadUrl("v1.2.0", $"{Asset}.sha256"),
                },
            },
        };

        Assert.Null(DesktopReleaseCatalog.AvailableUpdate(
            JsonSerializer.SerializeToUtf8Bytes(new[] { release }),
            "v1.1.0"));
    }

    [Fact]
    public void ReturnsNoUpdateForDevelopmentOrCurrentRelease()
    {
        var catalog = Catalog(
            Release("v1.2.0", "2026-09-20T00:00:00Z", assets: [Asset, $"{Asset}.sha256"]));

        Assert.Null(DesktopReleaseCatalog.AvailableUpdate(catalog, "development"));
        Assert.Null(DesktopReleaseCatalog.AvailableUpdate(catalog, "v1.2.0"));
    }

    [Fact]
    public void RejectsMalformedCatalogs()
    {
        Assert.Throws<DesktopReleaseCatalogException>(() =>
            DesktopReleaseCatalog.AvailableUpdate("{"u8, "v1.0.0"));
    }

    [Fact]
    public void AtomFallbackSelectsNewestReleaseAndBuildsTrustedUrls()
    {
        var feed = AtomFeed(
            AtomEntry("v1.2.0", "2026-09-20T00:00:00Z"),
            AtomEntry("v1.3.0", "2026-09-21T00:00:00Z"));

        var update = Assert.IsType<DesktopReleaseUpdate>(
            DesktopReleaseCatalog.AvailableUpdateFromAtom(feed, "v1.2.0"));

        Assert.Equal("v1.3.0", update.Tag);
        Assert.Equal(Asset, update.AssetName);
        Assert.Equal(DownloadUrl("v1.3.0", Asset), update.ArchiveUrl.AbsoluteUri);
        Assert.Equal(
            DownloadUrl("v1.3.0", $"{Asset}.sha256"),
            update.ChecksumUrl.AbsoluteUri);
    }

    [Fact]
    public void AtomFallbackDoesNotDowngradeNewerInstalledRelease()
    {
        var feed = AtomFeed(
            AtomEntry("v2.0.0", "2026-09-21T00:00:00Z"),
            AtomEntry("v1.9.0", "2026-09-20T00:00:00Z"));

        Assert.Null(DesktopReleaseCatalog.AvailableUpdateFromAtom(feed, "v2.0.0"));
    }

    [Fact]
    public void AtomFallbackIgnoresUntrustedAndUnsafeReleaseLinks()
    {
        var feed = AtomFeed(
            AtomEntry(
                "v9.0.0",
                "2026-09-22T00:00:00Z",
                "https://example.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/tag/v9.0.0"),
            AtomEntry(
                "unsafe",
                "2026-09-21T00:00:00Z",
                "https://github.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/tag/not-a-release"),
            AtomEntry("v1.2.0", "2026-09-20T00:00:00Z"));

        var update = Assert.IsType<DesktopReleaseUpdate>(
            DesktopReleaseCatalog.AvailableUpdateFromAtom(feed, "v1.1.0"));
        Assert.Equal("v1.2.0", update.Tag);
    }

    [Theory]
    [InlineData("<feed><entry>")]
    [InlineData("<!DOCTYPE feed [<!ENTITY xxe SYSTEM 'file:///etc/passwd'>]><feed>&xxe;</feed>")]
    public void AtomFallbackRejectsMalformedOrDtdFeeds(string feed)
    {
        Assert.Throws<DesktopReleaseCatalogException>(() =>
            DesktopReleaseCatalog.AvailableUpdateFromAtom(
                Encoding.UTF8.GetBytes(feed),
                "v1.0.0"));
    }

    private static byte[] Catalog(params object[] releases) =>
        JsonSerializer.SerializeToUtf8Bytes(releases);

    private static object Release(
        string tag,
        string publishedAt,
        bool draft = false,
        string[]? assets = null) => new
        {
            tag_name = tag,
            published_at = publishedAt,
            draft,
            assets = (assets ?? []).Select(name => new
            {
                name,
                browser_download_url = DownloadUrl(tag, name),
            }),
        };

    private static string DownloadUrl(string tag, string name) =>
        $"https://github.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/download/{tag}/{name}";

    private static byte[] AtomFeed(params string[] entries) => Encoding.UTF8.GetBytes($$"""
        <?xml version="1.0" encoding="UTF-8"?>
        <feed xmlns="http://www.w3.org/2005/Atom">
          {{string.Join("\n", entries)}}
        </feed>
        """);

    private static string AtomEntry(
        string tag,
        string updated,
        string? href = null) => $$"""
        <entry>
          <updated>{{updated}}</updated>
          <link rel="alternate" href="{{href ?? $"https://github.com/IPA-CyberLab/IPA-RS-HeteroNetwork/releases/tag/{tag}"}}" />
        </entry>
        """;
}
